include/internal/hashtable.h

```diff
@@ -0,0 +1,319 @@
+    v->value = data;                                                           \
+    return v;                                                                  \
+}                                                                              \
+                                                                               \
+pfx ossl_unused int ossl_ht_##name##_##vtype##_type(HT_VALUE *h)               \
+{                                                                              \
+    return h->type_id == name##_##vtype##_id;                                  \
+}
+
+#define DECLARE_HT_VALUE_TYPE_FNS(vtype, name)                                 \
+int ossl_ht_##name##_##vtype##_insert(HT *h, HT_KEY *key, vtype *data,         \
+                                      vtype **olddata);                        \
+vtype *ossl_ht_##name##_##vtype##_from_value(HT_VALUE *v);                     \
+vtype *ossl_unused ossl_ht_##name##_##vtype##_get(HT *h,                       \
+                                                  HT_KEY *key,                 \
+                                                  HT_VALUE **v);               \
+HT_VALUE *ossl_ht_##name##_##vtype##_to_value(vtype *data, HT_VALUE *v);       \
+int ossl_ht_##name##_##vtype##_type(HT_VALUE *h);                              \
+
+/*
+ * Helper function to construct case insensitive keys
+ */
+static inline void ossl_ht_strcase(char *tgt, const char *src, int len)
+{
+    int i;
+#if defined(CHARSET_EBCDIC) && !defined(CHARSET_EBCDIC_TEST)
+    const long int case_adjust = ~0x40;
+#else
+    const long int case_adjust = ~0x20;
+#endif
+
+    if (src == NULL)
+        return;
+
+    for (i = 0; src[i] != '\0' && i < len; i++)
+        tgt[i] = case_adjust & src[i];
+}
+
+/*
+ * Create a new hashtable
+ */
+HT *ossl_ht_new(HT_CONFIG *conf);
+
+/*
+ * Frees a hash table, potentially freeing all elements
+ */
+void ossl_ht_free(HT *htable);
+
+/*
+ * Empties a hash table, potentially freeing all elements
+ */
+int  ossl_ht_flush(HT *htable);
+
+/*
+ * Inserts an element to a hash table, optionally returning
+ * replaced data to caller
+ */
+int ossl_ht_insert(HT *htable, HT_KEY *key, HT_VALUE *data,
+                   HT_VALUE **olddata);
+
+/*
+ * Deletes a value from a hash table, based on key
+ * results provides information about delete status
+ */
+#define HT_DEL_FLAG_DELETED            0x1
+#define HT_DEL_FLAG_VALUE_NOT_RETURNED 0x2
+#define HT_DEL_FLAG_DEL_NOT_SUPPORTED  0x4
+HT_VALUE *ossl_ht_delete(HT *htable, HT_KEY *key, uint32_t *results);
+
+/*
+ * Returns number of elements in the hash table
+ */
+size_t ossl_ht_count(HT *htable);
+
+/*
+ * Iterates of the hash table calling cb for each element
+ */
+void ossl_ht_foreach(HT *htable, void (*cb)(HT_VALUE *obj, void *arg),
+                     void *arg);
+
+/*
+ * Variant of foreach, but aborts the loop when cb returns 0
+ */
+void ossl_ht_foreach_until(HT *htable, int (*cb)(HT_VALUE *obj, void *arg),
```

Comment:
Just have this function, there's no need for both.

Reply:
ok

Reply:
removed